### PR TITLE
core:libmbedtls: Add support for 128 bit keys to 3DES CMAC

### PR DESCRIPTION
**Pull request by uglym8**
Awhile ago I asked this question
https://github.com/OP-TEE/optee_os/issues/4443

I wrongly assumed that both libtomcrypt and libmdetls don't handle 128 bit 3DES keys correctly and that I have to copy K1 to K3 in order to get correct result for 3DES CMAC with this type of keys.

It turned out that only libmbedtls needs to be changed in order to work correctly with 3DES 128 bit keys.

The story with libtomcrypt implementation was amusing to say the least:

While implementing HW accelerator support for DES/3DES I made a mistake with handling of 128 bit 3DES keys in libtomcrypt (namely in des3_setup() - core/lib/libtomcrypt/src/ciphers/des.c)
In this function 128 bit keys get extended to 192 bit keys by copying the K1 into K3.

Unfortunately this wasn't caught by xtest due to way TEE_CipherInit GP function is implemented if libtomcrypt is used as core crypto library for OP-TEE.

Details:
In the xtest code (namely, 4003) there are two checks related to 128 bit 3DES keys:

-     for TEE_ALG_DES3_ECB_NOPAD algorithm
-     for TEE_ALG_DES3_CBC_NOPAD algorithm

Eventually after all preparation steps TEE_CipherInit() gets called.
To simplify things, let's just say that,  if libtomcrypt is used, it will eventually call another function before calling des3_setup():

    ltc_ecb_init() for  TEE_ALG_DES3_ECB_NOPAD algorithm

    ltc_cbc_init()  TEE_ALG_DES3_CBC_NOPAD algorithm

Both these function call get_des2_key() which converts 128 bit key to 192 key.

This basically means that if there's an error in handling 128 bit key in des3_setup() it won't be detected by tests which use TEE_Cipher()* functions because by the the time key gets to de3_setup() it will be already extended to 192 bits thus code path related to extending the key won't be taken.

I also prepared a PR with additional test to check 128 bit key handling for 3DES CMAC.

### CODE REVIEW

File: lib/libmbedtls/mbedtls/library/cmac.c [hunk 1]
```diff
@@ -235,6 +235,7 @@ int mbedtls_cipher_cmac_starts( mbedtls_cipher_context_t *ctx,
         case MBEDTLS_CIPHER_AES_128_ECB:
         case MBEDTLS_CIPHER_AES_192_ECB:
         case MBEDTLS_CIPHER_AES_256_ECB:
+        case MBEDTLS_CIPHER_DES_EDE_ECB:
```
Comment (jforissier) [inline lib/libmbedtls/mbedtls/library/cmac.c:238]: That doesn't look right. It seems this will allow DES in CMAC mode but I think only 3DES is valid for this mode. In any case this is a change to the internals of MBedTLS, so it would need to be a separate patch ("libmbedtls: blah blah") with the OP-TEE core specific changes added on top as you did for the rest of the PR ("core: libmbedtls: add support...").

File: lib/libmbedtls/core/cmac.c [hunk 2]
```diff
@@ -39,6 +39,17 @@ static TEE_Result mbed_cmac_init(struct crypto_mac_ctx *ctx,
...
+		/*
+		 * mbedTLS supports 128 bit keys for MBEDTLS_CIPHER_ID_DES
+		 * and not for MBEDTLS_CIPHER_ID_3DES.
+		 * That's why we have to change the cipher id here.
+		 */
+		if (c->cipher_id == MBEDTLS_CIPHER_ID_3DES) {
+			c->cipher_id = MBEDTLS_CIPHER_ID_DES;
+		}
+	}
+
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:52]: I'm not sure I understand all what is happening here, but frankly I don't like this much. You are pretending to use an algorithm when you are using a different one, right? Why not simply create a proper 192-bit key from the 128-bit one and be done with it?

**Review by jforissier — COMMENTED** [review]
Hi @uglym8, thanks for the patch, see my comments below.

**Comment by uglym8** [comment]
@jforissier thanks a lot for your review.

I also created a PR for upstream mbedTLS as well but, frankly speaking, I doubt that it will make it through as 3DES is really not of the top priority nowdays:)
https://github.com/ARMmbed/mbedtls/pull/4221

File: lib/libmbedtls/mbedtls/library/cmac.c (hunk 1 above)
Comment (uglym8) [reply lib/libmbedtls/mbedtls/library/cmac.c:238]: MBEDTLS_CIPHER_DES_EDE_ECB is the name of 3DES with 128 bit keys in mbedTLS.
Yes, I know sounds strange;)

File: lib/libmbedtls/core/cmac.c (hunk 2 above)
Comment (uglym8) [reply lib/libmbedtls/core/cmac.c:52]: Actually I had at least 4 ways how to do it:
1) the most straightforward one:
just extend 128 bit keys here to 192 bits and let it flow
2) mbedTLS in theory has a flag for algorithm that indicates that it support keys of variable length but using this approach will require more changes in mbedTLS core than any of the approaches described here.
3) don't setup cipher in crypto_cmac_alloc_ctx() in a case of 3DES cipher and setup it at mbed_cmac_init() step when key length is known.
This still requires the same changes as proposed in this very PR though.
4) this variant which seems as a good compromise to me as the required changes are minimal compared to (2) and (3), and at the same time we are not required to copy keys here (so we keep the same behaviour as in case if libtomcrypt  is used as a core crypto library)

Comment (jforissier) [reply lib/libmbedtls/core/cmac.c:52]: I'd like to see option (1) actually. I wouldn't be surprise it would be the simplest, easiest to understand and to maintain.

File: lib/libmbedtls/mbedtls/library/cmac.c (hunk 1 above)
Comment (jforissier) [reply lib/libmbedtls/mbedtls/library/cmac.c:238]: OK thanks for clarifying, in that case this whole change looks reasonable.

File: lib/libmbedtls/core/cmac.c (hunk 2 above)
Comment (jforissier) [reply lib/libmbedtls/core/cmac.c:52]: I'd prefer a single `if ()`:
```C
	if (len == 16 && c->cipher_id == MBEDTLS_CIPHER_ID_3DES) {
		/*
		 * mbedTLS supports 128 bit keys for MBEDTLS_CIPHER_ID_DES
		 * and not for MBEDTLS_CIPHER_ID_3DES.
		 * That's why we have to change the cipher id here.
		 */
		c->cipher_id = MBEDTLS_CIPHER_ID_DES;
	}
```

**Comment by jforissier** [comment]
With my minor stylistic comment addressed and the Travis (checkpatch) warning fixed, and also after splitting the mbedtls specific part into its own commit:
`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jforissier** [comment]
> @jforissier thanks a lot for your review.
> 
> I also created a PR for upstream mbedTLS as well but, frankly speaking, I doubt that it will make it through as 3DES is really not of the top priority nowdays:)
> [ARMmbed/mbedtls#4221](https://github.com/ARMmbed/mbedtls/pull/4221)

Well the change is simple enough to review so hopefully they merge it!

**Comment by jenswi-linaro** [comment]
It seems this review has stalled. I would be nice if we could get that last bit fixed so this can be merged.

**Comment by uglym8** [comment]
Sorry for keeping you waiting guys.
I'll get back to this in a very short time.
